### PR TITLE
Tap 1.2.1

### PR DESCRIPTION
**Pull request by jcsegovia**
New capabilities added:
-Tables sharing.
-Tables editing.
-Tables deleting.
-Tables upload using pytables & jobs results.
-Cross match.
-Data access.
-Datalink access.

### CODE REVIEW

**Comment by jcsegovia** [comment]
@bsipocz there is an error in Travis CI build, but I am not sure it is due the changes we have done:

`Exception occurred:
  File "/home/travis/build/astropy/astroquery/astropy_helpers/astropy_helpers/extern/automodapi/utils.py", line 82, in <listcomp>
    pkgitems = [(k, mod.__dict__[k]) for k in mod.__all__]
KeyError: 'Conf'
The full traceback has been saved in /tmp/sphinx-err-mm1t4pqo.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
`
Please, could you check it...?

**Comment by jcsegovia** [comment]
@keflavich @bsipocz it seems there is an exception when creating documentation. I do not know how to access to the log file that contains the error. In any case, I think it is not related to gaia nor tap modules.
Please, could you check it?

`Exception occurred:
  File "/home/travis/build/astropy/astroquery/astropy_helpers/astropy_helpers/extern/automodapi/utils.py", line 82, in <listcomp>
    pkgitems = [(k, mod.__dict__[k]) for k in mod.__all__]
KeyError: 'Conf'
The full traceback has been saved in /tmp/sphinx-err-dq1um9x7.log, ...`

**Comment by keflavich** [comment]
@jcsegovia I'll have a look, but I bet @bsipocz knows better.  Have you built successfully locally?

**Comment by jcsegovia** [comment]
@keflavich I do not create Travis builds locally. I have never used Travis before.
I usually develop my python module using an IDE and I test it using 'python setup.py install' in an specific environment, so I can use Anaconda with it (for instance).
I have done those tests locally, and my changes are working.

**Comment by pllim** [comment]
@jcsegovia , (tldr) if the error is coming from doc build, you can probably reproduce it by running `python setup.py build_sphinx`. Unit tests will not run into doc build errors.

p.s. I see that you are TAPping into the service's full potential. 😏

**Comment by keflavich** [comment]
I am able to reproduce the failure locally, so there is a real issue.

**Comment by keflavich** [comment]
The problem is that you commented out the entire configuration block in `astroquery/gaia/__init__.py`, but left `Conf` in `__all__`.  Did you intentionally remove the configuration block?

File: astroquery/gaia/__init__.py [hunk 1]
```diff
@@ -17,12 +17,13 @@
 
 from astroquery.utils.tap.core import TapPlus
 from astropy import config as _config
+from .core import Gaia, GaiaClass
 
 
+"""
```
Comment (keflavich) [inline astroquery/gaia/__init__.py]: here, you've commented out the whole `Conf` block

File: astroquery/gaia/__init__.py [hunk 2]
```diff
@@ -38,4 +39,6 @@ class Conf(_config.ConfigNamespace):
 
 from .core import Gaia, GaiaClass
 
+"""
+
 __all__ = ['Gaia', 'GaiaClass', 'Conf', 'conf']
```
Comment (keflavich) [inline astroquery/gaia/__init__.py]: but `Conf` is still used here, which is why the docs are failing

File: astroquery/gaia/core.py [hunk 3]
```diff
@@ -19,10 +19,25 @@
...
+
+    MAIN_GAIA_TABLE = _config.ConfigItem("gaiadr2.gaia_source",
+                                         "GAIA source data table")
+    MAIN_GAIA_TABLE_RA = _config.ConfigItem("ra",
+                                            "Name of RA parameter in table")
+    MAIN_GAIA_TABLE_DEC = _config.ConfigItem("dec",
+                                             "Name of Dec parameter in table")
+
+
+conf = Conf()
```
Comment (keflavich) [inline astroquery/gaia/core.py]: Why did you move the configuration here instead of leaving it in `__init__`?

**Comment by jcsegovia** [comment]
@pllim @keflavich  thanks! I was able to find the error by executing 'python setup.py build_sphinx'
I see no reason for moving conf into gaia core.py. I have moved it into __init__.py (original location).

File: astroquery/gaia/__init__.py (hunk 2 above)
Comment (jcsegovia) [reply astroquery/gaia/__init__.py]: Yes, you are right.

File: astroquery/gaia/core.py (hunk 3 above)
Comment (jcsegovia) [reply astroquery/gaia/core.py]: I will move conf to __init__.py.
Thanks.

**Comment by bsipocz** [comment]
Ohh, I'm sorry, I totally lost track of this. Let me do a review tomorrow (but given the size of the PR I suppose I could mostly do nitpicks, etc rather than anything thorough and trust that this works as it suppose to work).

File: astroquery/gaia/core.py [hunk 4]
```diff
@@ -549,5 +698,356 @@ def __getCoordInput(self, value, msg):
...
+            field name can be 'utype', 'ucd', 'flags' or 'indexed'
+            value is the new value this field of this column will take
+        verbose : bool, optional, default 'False'
+            flag to display information about the process
+        """
+        return self.__gaiatap.update_user_table(table_name,
+                                                list_of_changes,
+                                                verbose)
+
+    def set_ra_dec_columns(self, table_name=None,
```
Comment (keflavich) [inline astroquery/gaia/core.py]: @jcsegovia There are several functions just like this one which duplicate the functionality of `TapPlus`.  It looks like a more straightforward approach, which would avoid all of this duplicate code, would be to have `GaiaClass` inherit from `TapPlus`.  Is there a reason this is not done?  If you want to discuss this in more detail live (e.g., via slack), let me know, I can arrange a time.

Comment (jcsegovia) [reply astroquery/gaia/core.py]: @keflavich you are right. I think it is a good idea to extend TapPlus, so we can remove duplicated code. Thanks!

**Comment by smoh** [comment]
[On the last master merge commit](https://github.com/astropy/astroquery/pull/1266/commits/e18a4a0c8a5cf563401fc61c246aee968792fb5c): Isn't it a bad idea to merge master into this pull request branch and introduce irrelevant changes from MAST/TESS stuff?

**Comment by keflavich** [comment]
@smoh Merging master won't pull in any irrelevant changes because the diff against master is still zero.  There is preference in astropy to rebase rather than merge PRs, but it's not a very big deal.

File: docs/gaia/gaia.rst [hunk 5]
```diff
@@ -582,6 +591,297 @@ To perform a logout
...
+  Job '1539932326689O' created to upload table 'table_test_from_url'.
+
+Now, you can query your table as follows:
+
+.. code-block:: python
+
+  >>> full_qualified_table_name = 'user_<your_login_name>.table_test_from_url'
+  >>> query = 'select * from ' + full_qualified_table_name
+  >>> job = Gaia.launch_job(query=query)
+  >>> results = job.get_resultsjob)
```
[the remark this answers is not in the document]
Comment (smoh) [reply docs/gaia/gaia.rst]: I think there's 9 other same typos ;-).

**Comment by bsipocz** [comment]
> On the last master merge commit: Isn't it a bad idea to merge master into this pull request branch and introduce irrelevant changes from MAST/TESS stuff?

You're right, in case of conflicting changes it can definitely can mess up the diff. Right now it's not the case as the TESS changes happened on totally disjunct part of the package.  
But definitely would like to see rebases rather than merges for branch clarity, also maybe some squashing, but latter is not that crucial.

**Comment by bsipocz** [comment]
Thanks you @smoh. 

> If the maintainers or authors would rather let this out first, I wouldn't want to hold back.

I would prefer to have the cleanup rather than merge this as is. 

@jcsegovia - let us know otherwise, e.g. if you have a pressing deadline by this has to be deployed.

**Comment by smoh** [comment]
A couple of more thoughts from tap/core.py:Tap:

- Similar to `Tap.__parseUrl`, `Tap.__appendData` has no reference to self and has no external usage so the code can just be directly where it is call (only) once in `__launchJobMultipart`. As an aside, the goal of this code is to build `"key1=value1&key2=value2&..."` string. Have the authors considered using package like [requests](http://docs.python-requests.org/en/master/) which is already a requirement for astroquery and would make these sort of things unnecessary?

- I suggest to reduce the print statements and use logging for debugging. Many print statements (usually activated with `verbose=`) are seemingly written for debugging and may not be user-oriented. With logging it's easy to set the desired verbosity for the entire script or just one code block. Currently there are more than 3x print calls in utils/tap than other modules combined.

```
❯ grep "[^>]\sprint(" astroquery -r | grep 'utils/tap' | wc -l
     182
❯ grep "[^>]\sprint(" astroquery -r | grep -v 'utils/tap' | wc -l
      54
```

**Comment by jcsegovia** [comment]
@bsipocz about deadline: no we do not have a deadline. We can work on this changes before merging this PR.

File: docs/gaia/gaia.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (jcsegovia) [reply docs/gaia/gaia.rst]: You are right. This must be fixed.

File: astroquery/utils/tap/core.py [hunk 6]
```diff
@@ -67,34 +91,54 @@ def __init__(self, url=None, host=None, server_context=None,
...
+                          server_context=server_context,
+                          tap_context=tap_context,
+                          upload_context=upload_context,
+                          table_edit_context=table_edit_context,
+                          data_context=data_context,
+                          datalink_context=datalink_context,
+                          port=port,
+                          sslport=sslport)
+            self.__connHandler = tap
         # if connectionHandler is set, use it (useful for testing)
```
[the remark this answers is not in the document]
Comment (jcsegovia) [reply astroquery/utils/tap/core.py]: You are right. If connHandler is set, no need to create connections.
About arguments consistency, again, you are right. I did not include these checks because the code would grow with multiple if's and, as you have said, this is a quite internal initialization.

File: astroquery/utils/tap/core.py [hunk 7]
```diff
@@ -67,34 +91,54 @@ def __init__(self, url=None, host=None, server_context=None,
...
+                              host=host,
+                              server_context=server_context,
+                              tap_context=tap_context,
+                              upload_context=upload_context,
+                              table_edit_context=table_edit_context,
+                              data_context=data_context,
+                              datalink_context=datalink_context,
+                              port=port,
+                              sslport=sslport)
+                self.__connHandler = tap
```
[the remark this answers is not in the document]
Comment (jcsegovia) [reply astroquery/utils/tap/core.py]: Fine. Thanks.

File: astroquery/utils/tap/core.py [hunk 8]
```diff
@@ -113,6 +157,39 @@ def load_tables(self, verbose=False):
...
+            print(response.status, errMsg)
+            raise requests.exceptions.HTTPError(errMsg)
+            return None
+        print("Parsing table '"+str(table)+"'...")
+        tsp = TableSaxParser()
+        tsp.parseData(response)
+        print("Done.")
+        return tsp.get_table()
+
     def __load_tables(self, only_names=False, include_shared_tables=False,
```
[the remark this answers is not in the document]
Comment (jcsegovia) [reply astroquery/utils/tap/core.py]: You are right. __load_tables should be move to TapPlus. Thanks.

**Comment by jcsegovia** [comment]
@smoh thanks for your comments. Please, feel free to modify the code if you want. Probably, we are not going to modify the code until next Thursday. About requests, I did not know it was a requirement for astroquery: I did not test it and I am not sure whether we are going to find any problem or not (we use cookies and multipart forms in several places, we need to check that it works with requests).

**Comment by keflavich** [comment]
@jcsegovia requests is a great package, and astroquery would be a total mess without it.  It handles all of the heavy lifting for remote queries, including cookies and multipart form data.

`astroquery` wraps `request` in the `BaseQuery` class, where we've added a layer that caches the returned content from the remote service.   You can choose whether or not to use that layer; it is sometimes very helpful, e.g., when you expect the same query to always return the same data, but should not be activated when the returns are dynamic and changing.

...we probably need more developer-directed documentation on these base classes...

**Comment by jcsegovia** [comment]
@smoh revert done!

**Comment by jcsegovia** [comment]
@smoh related to TapPlus specific Gaia capabilities, I would like to clarify that TAP+ is used for several ESA archives. So I would not like to move TapPlus inside Gaia package. If utils.tap is not the right place, I would like to create something like utils.tapplus so we can put a common code inside.

**Comment by smoh** [comment]
@jcsegovia I see, thanks for letting me know.

@keflavich @bsipocz: Since there is another pull request #1288 depending on TapPlus and its API and possibly more coming from ESA ESDC, I think I better sign off my review here (the code as is before any of my commits). The code works as intended. Perhaps the suggested changes can be made in a later separate pull request.

Since none of the changes I introduced in the pull request to fork (which is incomplete) is here after the revert commit, the maintainers should feel free to merge. I don't know if it's best practice to still preserve commit history or force-delete; I'll just leave it up to you guys.

File: astroquery/utils/tap/model/job.py [hunk 9]
```diff
@@ -201,39 +304,164 @@ def wait_for_job_end(self, verbose=False):
         """
         currentResponse = None
         responseData = None
+        lphase = None
+        # execute job if not running
+        if self._phase == 'PENDING':
+            print("Job in PENDING phase, sending phase=RUN request.")
+            try:
+                self.start(verbose)
+            except:  # noqa: E722
```
Comment (keflavich) [inline astroquery/utils/tap/model/job.py]: better to do `except Exception as ex:` here, then `print(ex)` or similar below

File: astroquery/utils/tap/taputils.py [hunk 10]
```diff
@@ -85,3 +102,119 @@ def set_top_in_query(query, top):
...
+    -------
+    A jobid.
+    """
+    pos = location.rfind('/')+1
+    jobid = location[pos:]
+    return jobid
+
+
+def get_schema_name(full_qualified_table_name):
+    """Extracts the schema name form a full qualified table name.
```
Comment (keflavich) [inline astroquery/utils/tap/taputils.py]:
```suggestion
    """Extracts the schema name from a full qualified table name.
```

File: docs/gaia/gaia.rst [hunk 11]
```diff
@@ -582,6 +591,297 @@ To perform a logout
...
+
+
+2.2.1. Uploading table from url to user space
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code-block:: python
+
+  >>> from astroquery.gaia import Gaia
+  >>> Gaia.login()
+  >>> url = "http://tapvizier.u-strasbg.fr/TAPVizieR/tap/sync/?REQUEST=doQuery&lang=ADQL&FORMAT=votable&QUERY=select+*+from+TAP_SCHEMA.columns+where+table_name='II/336/apass9'"
```
Comment (keflavich) [inline docs/gaia/gaia.rst]: Are users supposed to know how to construct these big URLs?

File: docs/utils/tap.rst [hunk 12]
```diff
@@ -445,6 +445,15 @@ It is possible to use a file where the credentials are stored:
   >>> gaia.login(credentials_file='my_credentials_file')
 
 
+If you do not provide any parameters at all, a prompt will ask for user name and password.
```
Comment (keflavich) [inline docs/utils/tap.rst]: Is this whole document a duplicate of the above document?

File: astroquery/gaia/tests/DummyTapHandler.py [hunk 13]
```diff
@@ -222,3 +223,125 @@ def logout(self, verbose=False):
         self.__invokedMethod = 'logout'
         self.__parameters['verbose'] = verbose
         return None
+
+    def load_data(self, params_dict, output_file=None, verbose=False):
+        self.__invokedMethod = 'load_data'
+        self.__parameters['params_dict'] = params_dict
+        self.__parameters['output_file'] = output_file
+        self.__parameters['verbose'] = verbose
+        return None
```
[the remark this answers is not in the document]
Comment (jcsegovia) [reply astroquery/gaia/tests/DummyTapHandler.py]: You are right.

File: astroquery/gaia/tests/test_gaiatap.py [hunk 14]
```diff
@@ -483,15 +348,213 @@ def __check_results_column(self, results, columnName, description, unit,
...
+        responseResultsJob.set_message("OK")
+        jobDataFile = data_path('job_1.vot')
+        jobData = utils.read_file_content(jobDataFile)
+        responseResultsJob.set_data(method='GET',
+                                    context=None,
+                                    body=jobData,
+                                    headers=None)
+        req = "async/" + jobid + "/results/result"
+        connHandler.set_response(req, responseResultsJob)
+        query = "SELECT crossmatch_positional(\
```
[the remark this answers is not in the document]
Comment (jcsegovia) [reply astroquery/gaia/tests/test_gaiatap.py]: Rewritten.

File: astroquery/utils/tap/model/job.py [hunk 15]
```diff
@@ -63,13 +65,109 @@ def __init__(self, async_job, query=None, connhandler=None):
...
+    def set_phase(self, phase):
+        """Sets the job phase
+
+        Parameters
+        ----------
+        phase : str, mandatory
+            job phase
+        """
+        if self.is_finished():
+            raise ValueError("Cannot assign a pahse when a job is finished")
```
[the remark this answers is not in the document]
Comment (jcsegovia) [reply astroquery/utils/tap/model/job.py]: OK

File: astroquery/utils/tap/model/job.py [hunk 16]
```diff
@@ -85,11 +183,13 @@ def get_phase(self, update=False):
...
             phase_request = "async/"+str(self.jobid)+"/phase"
-            response = self.connHandler.execute_get(phase_request)
+            response = self.connHandler.execute_tapget(phase_request)
 
             self.__last_phase_response_status = response.status
             if response.status != 200:
-                raise Exception(response.reason)
+                errMsg = taputils.get_http_response_error(response)
+                print(response.status, errMsg)
+                raise Exception(errMsg)
```
[the remark this answers is not in the document]
Comment (jcsegovia) [reply astroquery/utils/tap/model/job.py]: I am not sure I like to use raise_for_status... taputils tries to parse the response in order to extract the suitable error message from server side.

File: astroquery/utils/tap/core.py [hunk 17]
```diff
@@ -113,6 +157,39 @@ def load_tables(self, verbose=False):
...
+        if verbose:
+            print(response.status, response.reason)
+        isError = self.__connHandler.check_launch_response_status(response,
+                                                                  verbose,
+                                                                  200)
+        if isError:
+            errMsg = taputils.get_http_response_error(response)
+            print(response.status, errMsg)
+            raise requests.exceptions.HTTPError(errMsg)
+            return None
```
[the remark this answers is not in the document]
Comment (jcsegovia) [reply astroquery/utils/tap/core.py]: OK

File: astroquery/utils/tap/core.py [hunk 18]
```diff
@@ -143,17 +220,20 @@ def __load_tables(self, only_names=False, include_shared_tables=False,
...
         isError = self.__connHandler.check_launch_response_status(response,
                                                                   verbose,
                                                                   200)
         if isError:
-            print(response.status, response.reason)
-            raise requests.exceptions.HTTPError(response.reason)
+            errMsg = taputils.get_http_response_error(response)
+            print(response.status, errMsg)
+            raise requests.exceptions.HTTPError(errMsg)
             return None
```
[the remark this answers is not in the document]
Comment (jcsegovia) [reply astroquery/utils/tap/core.py]: OK

File: astroquery/utils/tap/core.py [hunk 19]
```diff
@@ -182,7 +262,8 @@ def launch_job(self, query, name=None, output_file=None,
             if True, the results are saved in a file instead of using memory
         upload_resource: str, optional, default None
             resource to be uploaded to UPLOAD_SCHEMA
-        upload_table_name: str, required if uploadResource is provided, default None
+        upload_table_name: str, required if upload_resource is
+            provided, default None
```
[the remark this answers is not in the document]
Comment (jcsegovia) [reply astroquery/utils/tap/core.py]: OK. Rewritten.

File: astroquery/utils/tap/core.py [hunk 20]
```diff
@@ -217,28 +299,39 @@ def launch_job(self, query, name=None, output_file=None,
                 response.getheaders(),
                 "location")
             if location is None:
-                raise requests.exceptions.HTTPError("No location found after redirection was received (303)")
+                raise requests.exceptions.HTTPError("No location found " +
+                                                    "after redirection was " +
```
[the remark this answers is not in the document]
Comment (jcsegovia) [reply astroquery/utils/tap/core.py]: OK

File: astroquery/utils/tap/core.py [hunk 21]
```diff
@@ -217,28 +299,39 @@ def launch_job(self, query, name=None, output_file=None,
...
         job.outputFile = suitableOutputFile
         job.parameters['format'] = output_format
         job.set_response_status(response.status, response.reason)
+        job.set_phase('PENDING')
         if isError:
-            job.set_failed(True)
+            job.failed = True
+            job.set_phase('ERROR')
+            responseBytes = response.read()
+            responseStr = responseBytes.decode('utf-8')
```
[the remark this answers is not in the document]
Comment (jcsegovia) [reply astroquery/utils/tap/core.py]: In fact, this is not required as the error handler (__connHandler.dump_to_file) works with response directly.

File: astroquery/utils/tap/core.py [hunk 22]
```diff
@@ -440,49 +556,65 @@ def save_results(self, job, verbose=False):
...
+            f = open(fh.name, "r")
+            chunk = f.read()
+            f.close()
+            os.unlink(fh.name)
+            name = 'pytable'
+            args['format'] = 'votable'
+        else:
+            f = open(uploadResource, "r")
+            chunk = f.read()
+            f.close()
```
[the remark this answers is not in the document]
Comment (jcsegovia) [reply astroquery/utils/tap/core.py]: OK

File: docs/gaia/gaia.rst (hunk 11 above)
Comment (jcsegovia) [reply docs/gaia/gaia.rst]: This is an example of an external URL.
Yes, it is supposed the user knows external URLs where a table can be found.
When I created this example, I decided to use an external URL that can provide real data.
If I use something like "http://tapvizier.u-strasbg.fr/TAPVizieR/tap/async/123456789/results/result", probably will not work (because that job does not exist) and the user can think that the example does not work.

File: astroquery/utils/tap/taputils.py (hunk 10 above)
Comment (jcsegovia) [reply astroquery/utils/tap/taputils.py]: OK

File: astroquery/utils/tap/model/job.py (hunk 9 above)
Comment (jcsegovia) [reply astroquery/utils/tap/model/job.py]: OK

File: astroquery/utils/tap/model/job.py [hunk 23]
```diff
@@ -63,13 +65,109 @@ def __init__(self, async_job, query=None, connhandler=None):
...
+                                                        data=data,
+                                                        verbose=verbose)
+            if verbose:
+                print(response.status, response.reason)
+                print(response.getheaders())
+            self.__last_phase_response_status = response.status
+            if response.status != 200:
+                errMsg = taputils.get_http_response_error(response)
+                print(response.status, errMsg)
+                raise Exception(errMsg)
```
[the remark this answers is not in the document]
Comment (jcsegovia) [reply astroquery/utils/tap/model/job.py]: See my comment above.

File: astroquery/utils/tap/model/job.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (keflavich) [reply astroquery/utils/tap/model/job.py]: Could you point me to where this is done?  I would think that the error message is the same either way.

I strongly recommend using something other than generic `Exception`s; in this case, the errors should almost certainly be at least `HTTPError`s

File: docs/gaia/gaia.rst (hunk 11 above)
Comment (keflavich) [reply docs/gaia/gaia.rst]: Alright. It might help to add in the documentation where that URL came from.

File: docs/utils/tap.rst (hunk 12 above)
Comment (keflavich) [reply docs/utils/tap.rst]: Yes, I think that's better when the code is _nearly_ the same.  But I don't feel strongly about this.

File: astroquery/utils/tap/model/job.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (jcsegovia) [reply astroquery/utils/tap/model/job.py]: taputils.get_http_response_error tries to find the error message in a VOTable response (field INFO, name 'QUERY_STATUS') or in an httml response (module taputils.py, line 107).
About avoiding generic Exception, I agree, I have changed Exeption by HTTPError.

File: docs/gaia/gaia.rst (hunk 11 above)
Comment (jcsegovia) [reply docs/gaia/gaia.rst]: OK. Doc. updated.

File: astroquery/utils/tap/model/job.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (keflavich) [reply astroquery/utils/tap/model/job.py]: Ahh, ok, this is for errors that are _not_ http errors.  Sounds good.

File: astroquery/utils/tap/core.py [hunk 24]
```diff
@@ -217,28 +291,40 @@ def launch_job(self, query, name=None, output_file=None,
...
+            response = self.__connHandler.execute_tapget(subcontext,
+                                                         verbose=verbose)
         job = Job(async_job=False, query=query, connhandler=self.__connHandler)
         isError = self.__connHandler.check_launch_response_status(response,
                                                                   verbose,
-                                                                  200)
+                                                                  200,
+                                                                  False)
+        headers = response.getheaders()
         suitableOutputFile = self.__getSuitableOutputFile(False,
```
Comment (emeraldTree24) [inline astroquery/utils/tap/core.py]: To add support for tsv format in the tapconn.py file in the function get_suitable_extension_by_format() could you add if tsv return .tsv and the in function get_suitable_extension() could you add if tab-separated-values return .tsv

**Comment by bsipocz** [comment]
This needs a rebase now.

**Comment by bsipocz** [comment]
I plan to open a follow-up PR that resurrects some of the cleanup from https://github.com/esdc-esac-esa-int/astroquery/pull/1, hopefully in the next couple of days, but feel free to open your own follow-ups as needed.
